pkg/services/auth/auth.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	publicKeyBytes := extractPublicKey(secret)
+
+	if err := a.gitProvider.UploadDeployKey(owner, repoName, publicKeyBytes); err != nil {
+		return fmt.Errorf("error uploading deploy key: %w", err)
+	}
+
+	if err := a.storeDeployKey(ctx, secret); err != nil {
+		return fmt.Errorf("error storing deploy key: %w", err)
+	}
+
+	a.logger.Println("Deploy key generated and uploaded to git provider")
+	// Set the gitClient to use the newly generated deploy key
+	a.gitClient = git.New(deployKey)
+
+	return nil
+
+}
+
+var ErrNoDeployKeysSetup = errors.New("the git client has not been intialized. Run SetupDeployKeys() on this object first.")
+
+// GitClient returns an instrumented git client that will use the deploy key any git operations.
+// SetupDeployKeys MUST be called before this, else an error will be returned.
+func (a *AuthService) GitClient() (git.Git, error) {
+	if a.gitClient == nil {
+		return nil, ErrNoDeployKeysSetup
+	}
+
+	return a.gitClient, nil
+}
+
+// GitProvider returns a GitProvider that is already constructed with access tokens.
+// This can be used for GitProvider API operations, like opening pull requests or adding deploy keys.
+func (a *AuthService) GitProvider() gitproviders.GitProvider {
+	return a.gitProvider
+}
+
+// Generates an ssh keypair for upload to the Git Provider and for use in a git.Git client.
+func (a *AuthService) generateDeployKey(targetName string, secretName types.NamespacedName, repo gitproviders.NormalizedRepoURL) (*ssh.PublicKeys, *corev1.Secret, error) {
+	secret, err := a.createKeyPairSecret(targetName, secretName, repo)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not create key-pair secret: %w", err)
+	}
+
+	privKeyBytes := extractPrivateKey(secret)
+
+	deployKey, err := makePublicKey(privKeyBytes)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating new public keys: %w", err)
+	}
+
+	return deployKey, secret, nil
+
+}
+
+// Wrapper to abstract how the key is stored in case we want to change this later.
+func (a *AuthService) storeDeployKey(ctx context.Context, secret *corev1.Secret) error {
+	if err := a.k8sClient.Create(ctx, secret); err != nil {
+		return fmt.Errorf("could not store secret: %w", err)
+	}
+
+	return nil
+}
+
+// Wrapper to abstract how a key is fetched.
+func (a *AuthService) retreiveDeployKey(ctx context.Context, name types.NamespacedName) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	if err := a.k8sClient.Get(ctx, name, secret); err != nil {
+		return nil, fmt.Errorf("error getting deploy key secret: %w", err)
+	}
+
+	return secret, nil
+}
+
+// Uses flux to create a ssh key pair secret.
+func (a *AuthService) createKeyPairSecret(targetName string, name types.NamespacedName, repo gitproviders.NormalizedRepoURL) (*corev1.Secret, error) {
+	secretname := gitproviders.CreateAppSecretName(targetName, repo)
+	secretData, err := a.fluxClient.CreateSecretGit(secretname, repo.String(), name.Namespace)
```

Comment:
How do you feel about refactoring the fluxClient to return the `corev1.Secret` instead of `[]byte` from `CreateSecretGit`?  It is called in two places, here and `Add`.  In both use-cases it is unmarshaled immediately into the `*corev1.Secret`.  This will clean up the code a lot and will help set a precedent to move away from `[]byte`.

Reply:
That would be great, but out of scope for this PR. Lots of other code will have to change.

Reply:
Bah, I missed in the other location that the secret is then passed to kube.Apply which takes `[]byte`.  Yeah, pull that thread and it is bigger than I thought it would be.